lib/scholar/metrics/clustering.ex

```diff
@@ -0,0 +1,155 @@
+defmodule Scholar.Metrics.Clustering do
+  @moduledoc """
+  Metrics related to clustering algorithms.
+  """
+
+  import Nx.Defn
+
+  defnp inner_dist(x, labels) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    expanded_points =
+      Nx.tile(x, [1, num_samples]) |> Nx.reshape({num_samples, num_samples, num_features})
+
+    labels_vectorize = Nx.broadcast(labels, {num_samples, num_samples}) |> Nx.transpose()
+
+    mask_zero_one =
+      labels_vectorize ==
+        Nx.tile(labels, [1, num_samples]) |> Nx.reshape({num_samples, num_samples})
+
+    denominator = Nx.sum(mask_zero_one, axes: [1]) |> Nx.reshape({num_samples, 1}) |> Nx.squeeze()
+    indices_to_zero = Nx.equal(denominator, 1)
+    denominator = Nx.select(Nx.greater(denominator, 1), denominator - 1, 1)
+
+    mask =
+      mask_zero_one
+      |> Nx.select(
+        Nx.tile(Nx.iota({num_samples}), [num_samples, 1]),
+        Nx.broadcast(Nx.iota({num_samples, 1}), {num_samples, num_samples})
+      )
+
+    points = Nx.take(x, mask)
+
+    {Nx.sum(
+       Scholar.Metrics.Distance.euclidean(
+         expanded_points,
+         points,
+         axes: [2]
+       ),
+       axes: [1]
+     ) / denominator, indices_to_zero}
+  end
+
+  defnp outer_dist(x, labels, opts \\ []) do
+    num_clusters = opts[:num_clusters]
+    {num_samples, num_features} = Nx.shape(x)
+    inf = Nx.Constants.infinity()
+
+    x_a =
+      x
+      |> Nx.reshape({1, num_samples, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    x_b =
+      x
+      |> Nx.reshape({num_samples, 1, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    pairwise_dist = Scholar.Metrics.Distance.euclidean(x_a, x_b, axes: [2])
+    labels_vectorize = Nx.tile(labels, [num_clusters, 1])
+    Nx.iota({num_clusters, num_samples}, axis: 0)
+    mask = labels_vectorize == Nx.iota({num_clusters, num_samples}, axis: 0)
+    denominator = mask |> Nx.sum(axes: [1]) |> Nx.broadcast({num_samples, num_clusters})
+    results = Nx.dot(mask, pairwise_dist) |> Nx.transpose()
+    results = Nx.select(Nx.transpose(mask), inf, results)
+    results = results / denominator
+    Nx.reduce_min(results, axes: [1])
```

Comment:
We can avoid the transposes and rely more on automatic broadcasting, something like this:

```suggestion
    membership_mask = Nx.reshape(labels, {num_samples, 1}) == Nx.iota({1, num_clusters})
    cluster_size = mask |> Nx.sum(axes: [0]) |> Nx.reshape({1, num_clusters})
    dist_in_cluster = Nx.dot(pairwise_dist, membership_mask)
    mean_dist_in_cluster = dist_in_cluster / cluster_size
    
    membership_mask
    |> Nx.select(inf, mean_dist_in_cluster)
    |> Nx.reduce_min(axes: [1])
```